gt/main.c

```diff
@@ -757,6 +841,124 @@ alloc_and_fill_notify_pkt(unsigned int socket, struct ggu_policy *policy,
+static void
+print_unsent_policy(struct ggu_policy *policy)
+{
+	int ret;
+	char err_msg[1024];
+
+	if (policy->state == GK_REQUEST) {
+		ret = snprintf(err_msg, sizeof(err_msg),
+			"gt: failed to send out the notification to Gatekeeper with policy decision [state: %hhu]",
+			policy->state);
+	} else if (policy->state == GK_GRANTED) {
+		ret = snprintf(err_msg, sizeof(err_msg),
+			"gt: failed to send out the notification to Gatekeeper with policy decision [state: %hhu, tx_rate_kb_sec: %u, cap_expire_sec: %u, next_renewal_ms: %u, renewal_step_ms: %u]",
+			policy->state, policy->params.u.granted.tx_rate_kb_sec,
+			policy->params.u.granted.cap_expire_sec,
+			policy->params.u.granted.next_renewal_ms,
+			policy->params.u.granted.renewal_step_ms);
+	} else if (policy->state == GK_DECLINED) {
+		ret = snprintf(err_msg, sizeof(err_msg),
+			"gt: failed to send out the notification to Gatekeeper with policy decision [state: %hhu, expire_sec: %u]",
+			policy->state,
+			policy->params.u.declined.expire_sec);
+	} else {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"gt: unknown policy decision at %s, there is a bug in the Lua policy!\n",
+			__func__);
+		return;
+	}
+
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(err_msg));
+	print_flow_err_msg(&policy->flow, err_msg);
+}
+
+/*
+ * Use the @dr to notify the GK
+ * about the punishment policies on declined flows
+ * with fragmented packets.
+ */
+static void 
+process_death_row(int socket_id, uint8_t port, uint16_t tx_queue,
+	struct rte_ip_frag_death_row *death_row,
+	struct gt_instance *instance, struct gt_config *gt_conf)
+{
+	uint32_t i;
+
+	for (i = 0; i < death_row->cnt; i++) {
+		struct gt_packet_headers pkt_info;
+		struct ggu_policy policy;
+		struct rte_mbuf *notify_pkt;
+		int ret = gt_parse_incoming_pkt(death_row->row[i], &pkt_info);
+		if (ret < 0) {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gt: failed to parse the fragments at %s, and the packet doesn't trigger any policy consultation at all!\n",
+				__func__);
+			rte_pktmbuf_free(death_row->row[i]);
+			continue;
+		}
+
+		/*
+		 * Given the gravity of the issue,
+		 * we must send a decline to the gatekeeper server
+		 * to expire in 10 minutes and log our failsafe action here.
+		 * Otherwise, a misconfigured grantor server can put
+		 * a large deployment at risk.
+		 */
+		ret = lookup_frag_punish_policy_decision(
+			&pkt_info, &policy, instance);
+		if (ret < 0) {
+			policy.state = GK_DECLINED;
+			policy.params.u.declined.expire_sec = 600;
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gt: failed to lookup the punishment policy for the packet fragment! Our failsafe action is to decline the flow for 10 minutes!\n");
+		}
+
+		/*
+		 * TODO Reply in a batch.
+		 * Reply the policy decision to GK-GT unit.
+		 */
+		notify_pkt = alloc_and_fill_notify_pkt(
+			socket_id, &policy, &pkt_info, gt_conf);
+		if (notify_pkt == NULL)
+			print_unsent_policy(&policy);
+
+		if (notify_pkt != NULL && rte_eth_tx_burst(
+				port, tx_queue, &notify_pkt, 1) != 1) {
```

Comment:
Style: turn this `if` test into an `else if` in the previous `if` statement; remember to drop `notify_pkt != NULL &&` after the change.